### DISCUSSION TITLE
Analyze ignore reader-macro #_ the same as comments

### DISCUSSION THREAD

**Discussion by snoe** [clj-kondo/clj-kondo#1236] [Ideas]
In our code bases, we often use `#_` in place of comments. It would be nice to analyze these the same as forms under `comment`

In this example, the `inc` inside the comment has analysis data, but the one behind the `#_` has nothing.
<img width="214" alt="image" src="https://user-images.githubusercontent.com/105012/112566044-e1a7fa80-8da3-11eb-9eee-b785b91d5068.png">

**Comment by borkdude** [comment]
The problem with treating `#_` as `comment` is that often `#_` is used to comment out arguments or key/vals in a map. It has different semantics as the reader is skipping over these fragments, while the `comment` macro always returns `nil`. So we can't simply treat `#_` in the same way as `comment`. Why not just use `comment`?

**Reply by snoe** [reply]
Well, lots of reasons... key/vals is one, it doesn't change indentation, you can comment without wrapping/unwrapping forms, you can comment out a number of forms from a single line with multiple `#_#_`s and like you say it doesn't eval to anything in the way that comment does. So it's closer to `;` while requiring the ignored forms to be readable. Mainly though, because that's not what people use, and I'd like to accommodate that.

So from a return value perspective, you're right it can't be treated exactly the same, my point is that the internals of the things being commented should be analyzed, as it's another way people comment out code.

**Reply by borkdude** [reply]
The example you give above seems to be a top-level invocation. Can you give a convincing example where you use `#_` inline, while you would still like to see warnings/get analysis?

**Reply by yuhan0** [reply]
Relevant issue: #929
Syntax errors in `#_` which prevent the file from being read should be linted, although I'm not convinced that any other analysis should take place.
E.g. one might want to `#_` an outdated var definition without getting redefined-var or mismatching arity warnings, something that does currently happen for `comment` forms.

**Reply by snoe** [reply]
Ok so you want to stop some function in a chain from being called while you use the repl. 

But then if they say find-references for `send-email!` (or worse carve and see `send-email!` isn't being used) then the user is working with an incomplete picture.

``` clojure
(defn do-notification [user alert]
  (-> alert
      (persist!)
      #_ (send-email! user)
      (send-websocket! user)))
```

Say you're refactoring some function and want to ignore a second call-site, and you decide to rename `refactoring` well it's still useful (and, I think, expected) for the ignored form to be renamed, same as in the comment.

```clojure
(defn a []
  (let [x (refactoring 123)]
    x))

(defn b []
  (let [x {:expected :return} #_(refactoring 456)]
    (something x)))

(comment
 (refactoring 789))
```

Why wouldn't you want to analyze these forms? 

The way I see people on my team work:
- `;` used to comment out things that aren't supposed to run
- `comment` to wrap repl execution mainly
- `#_` used to temporarily comment things inline.

**Reply by borkdude** [reply]
@snoe After removal of `send-email!` or renaming `refactoring`, these symbols show up as unresolved when you decide to "uncomment". This is true for both `;` and `#_`.

**Reply by snoe** [reply]
Ok, I'm less concerned about linting, and making sure it shows up in analysis so tools can make their own determination.

**Reply by borkdude** [reply]
@snoe I'm starting to see your point, thanks for being persistent. It's a bit painful to start supporting this now, but maybe it can be done. 

Currently the `#_` forms are thrown away during parsing, like other white-space related things. So to support your use case we have to:

1) make this an option to not throw away these nodes

But considering #929 we might already need the nodes anyway, for light-weight linting of unresolved aliases for example.
 
2) we must check, I think in `analyze-expression**` if something is of this node type and then lint it accordingly.

Feel free to make an experimental branch with this, if you would like. If it will work out without too much pain and risk of (performance + correctness) regressions, we can create a proper issue + PR out of this.
